### PR TITLE
Fix index selectivity estimate issues with RocksDB engine

### PR DESCRIPTION
**Pull request by danielhlarkin**

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBBackgroundThread.cpp [hunk 1]
```diff
@@ -58,7 +58,18 @@ void RocksDBBackgroundThread::run() {
...
+          LOG_TOPIC(WARN, Logger::ENGINES)
+            << "background settings sync failed: " << res.errorMessage();
+        }
+
+        double end = TRI_microtime();
+        if ((end - start) > 0.5) {
+          LOG_TOPIC(WARN, Logger::ENGINES)
+            << "slow background settings sync: " << (end - start) << "sec";
+        }
       }
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBBackgroundThread.cpp]: nice i like this. May help to debug :+1:

File: arangod/RocksDBEngine/RocksDBRecoveryManager.cpp [hunk 2]
```diff
@@ -307,10 +306,25 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
       if (hash != 0) {
         uint64_t objectId = RocksDBKey::objectId(key);
         auto est = findEstimator(objectId);
-        if (est.first != nullptr && est.second < currentSeqNum) {
+        if (est != nullptr && est->commitSeq() < currentSeqNum) {
+          /*LOG_TOPIC(TRACE, Logger::ENGINES)
+            << "recovery adding hash'" << hash
+            << "' to estimator for index with objectId'" << objectId << "'";*/
+
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBRecoveryManager.cpp]: Can we either activate or remove this comment here? Both are fine wtih me.

File: arangod/RocksDBEngine/RocksDBRecoveryManager.cpp [hunk 3]
```diff
@@ -307,10 +306,25 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
...
+          if (est.first == nullptr) {
+            LOG_TOPIC(TRACE, Logger::ENGINES)
+              << "not tracking estimator for index with objectId' "
+              << objectId << "', estimator not found";
+          } else {
+            LOG_TOPIC(TRACE, Logger::ENGINES)
+              << "not tracking estimator for index with objectId' "
+              << objectId << "' (synced at " << est.second << " >= "
+              << currentSeqNum << ")";
+          }
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBRecoveryManager.cpp]: Can we activate the trace output here?
Should help for debugging if index estimates are off.

File: arangod/RocksDBEngine/RocksDBRecoveryManager.cpp [hunk 4]
```diff
@@ -348,10 +362,25 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
...
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBRecoveryManager.cpp]: Can we activate the logging here for debugging purposes as well?

File: arangod/RocksDBEngine/RocksDBRecoveryManager.cpp [hunk 5]
```diff
@@ -348,10 +362,25 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
       if (hash != 0) {
         uint64_t objectId = RocksDBKey::objectId(key);
         auto est = findEstimator(objectId);
-        if (est.first != nullptr && est.second < currentSeqNum) {
+        if (est != nullptr && est->commitSeq() < currentSeqNum) {
+          /*LOG_TOPIC(TRACE, Logger::ENGINES)
+            << "recovery removing hash'" << hash
+            << "' from estimator for index with objectId'" << objectId << "'";*/
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBRecoveryManager.cpp]: Can we active the logging for debugging here?

File: tests/RocksDBEngine/IndexEstimatorTest.cpp [hunk 6]
```diff
@@ -91,27 +91,29 @@ SECTION("test_serialize_deserialize") {
...
 
   // After serialisation => deserialisation
   // both estimates have to be identical
   CHECK(est.nrUsed() == copy.nrUsed());
   CHECK(est.nrCuckood() == copy.nrCuckood());
   CHECK(est.computeEstimate() == copy.computeEstimate());
+  CHECK(seq == copy.commitSeq());
 
   // Now let us remove the same elements in both
   bool coin = false;
```
Comment (mchacki) [inline tests/RocksDBEngine/IndexEstimatorTest.cpp]: Can we add some Catch Tests to proof the blocker logic on this level already?

**Comment by danielhlarkin** [comment]
@mchacki Agreed. I'll take care of adding tests and enabling/removing commented out logging statements as appropriate. Also probably need to add some documentation.

**Pull request merged**